plugins/spi/SPIOutput.cpp

```diff
@@ -486,6 +502,135 @@ uint8_t SPIOutput::P9813CreateFlag(uint8_t red, uint8_t green, uint8_t blue) {
+    unsigned int spi_offset = APA102_START_FRAME_BYTES +
+                              (i * APA102_SPI_BYTES_PER_PIXEL);
+    // set pixel data
+    // first Byte contains:
+    // 3bit start mark (111) + 5bit GlobalBrightnes
+    // set GlobalBrightnes fixed to 31 --> that reduces flickering
+    // that can be written as 0xE0 & 0x1F
+    output[spi_offset + 0] = 0xFF;
+    // only write pixel data if new data from buffer available:
+    if ((buffer.Size() - offset) >= APA102_SLOTS_PER_PIXEL) {
+      // Convert RGB to APA102 Pixel
+      uint8_t r = 0;
+      uint8_t b = 0;
+      uint8_t g = 0;
+      // read rgb values
+      r = buffer.Get(offset + 0);
+      g = buffer.Get(offset + 1);
+      b = buffer.Get(offset + 2);
+      // set rgb values in output
+      // skip spi_offset + 0 (is already set)
+      output[spi_offset + 1] = b;
+      output[spi_offset + 2] = g;
+      output[spi_offset + 3] = r;
+    }
+  }
+  m_backend->Commit(m_output_number);
+}
+
+void SPIOutput::CombinedAPA102Control(const DmxBuffer &buffer) {
+  // for Protocol details see IndividualAPA102Control
+
+  const uint8_t latch_bytes = CalculateAPA102LatchBytes(m_pixel_count);
+  const unsigned int first_slot = m_start_address - 1;  // 0 offset
+
+  // check if enough data is there.
+  if (buffer.Size() - first_slot < APA102_SLOTS_PER_PIXEL) {
+    OLA_INFO << "Insufficient DMX data, required " << APA102_SLOTS_PER_PIXEL
+             << ", got " << buffer.Size() - first_slot;
+    return;
+  }
+
+  // get data for entire string length
+  const unsigned int output_length = (m_pixel_count + 1) *
+                                      APA102_SPI_BYTES_PER_PIXEL;
+  uint8_t *output = m_backend->Checkout(m_output_number, output_length,
+                                        latch_bytes);
+  // only update spi data if possible
+  if (!output)
+    return;
+
+  // create Pixel Data
+  uint8_t pixel_data[APA102_SPI_BYTES_PER_PIXEL];
+  pixel_data[0] = 0xff;
+  pixel_data[1] = buffer.Get(first_slot + 2);  // Get Blue
+  pixel_data[2] = buffer.Get(first_slot + 1);  // Get Green
+  pixel_data[3] = buffer.Get(first_slot);  // Get Red
+
+  // set all pixel to same value
+  for (unsigned int i = 0; i < m_pixel_count; i++) {
+    unsigned int spi_offset = APA102_START_FRAME_BYTES +
+                              (i * APA102_SPI_BYTES_PER_PIXEL);
+    memcpy(&output[spi_offset], pixel_data,
+           APA102_SPI_BYTES_PER_PIXEL);
+  }
+
+  // write output back...
+  m_backend->Commit(m_output_number);
+}
+
+/**
+ * Calculate Latch Bytes for APA102:
+ * minimal use half the pixel count bits 
+ * round up to next full byte count.
+ * datasheet says endframe should consist of 4 bytes - 
+ * but thats only valid for up to 64 pixels/leds. (4Byte*8Bit*2=64)
+ *
+ * the function is valid up to 4080 pixels. (255*8*2)
+ * ( otherwise the return type must be changed to uint16_t)
+ */
+uint8_t SPIOutput::CalculateAPA102LatchBytes(unsigned int m_pixel_count) {
```

Comment:
m_pixel_count shadows the member variable with the same name. Call this pixel_count and make the method static.